### ISSUE TITLE
curations: version range not working for 4 digit versions

### ISSUE THREAD

**Issue by mawl** [oss-review-toolkit/ort#6153]
Hey,

Concerning curations, it seems that ivy-style version ranges doesn't work for four digit versions like 1.1.1.3.

ivy-style version matcher introduction: https://docs.gradle.org/current/userguide/single_versions.html

### single version 1.1.1.3
NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.3
```
    - metadata:
        id: "NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.3"
      curations:
      - base: {}
        curation:
          comment: "https://go.microsoft.com/fwlink/?linkid=865381 contains MIT License"
          declared_license_mapping:
            https://go.microsoft.com/fwlink/?linkid=865381: "MIT"
```

### prefix matcher 1.1.1.+
NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.+
```
    - metadata:
        id: "NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.3"
      curations:
      - base: {}
        curation:
          comment: "https://go.microsoft.com/fwlink/?linkid=865381 contains MIT License"
          declared_license_mapping:
            https://go.microsoft.com/fwlink/?linkid=865381: "MIT"
```

### prefix matcher 1.+
NuGet::Xamarin.Android.Arch.Core.Common:1.+
```
    - metadata:
        id: "NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.3"
      curations:
      - base: {}
        curation:
          comment: "https://go.microsoft.com/fwlink/?linkid=865381 contains MIT License"
          declared_license_mapping:
            https://go.microsoft.com/fwlink/?linkid=865381: "MIT"
```

### version range [1.1.1,)
NuGet::Xamarin.Android.Arch.Core.Common:\[1.1.1,)
```
    - metadata:
        id: "NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.3"
      curations: []
```

### version range [1.1.1.1,)
NuGet::Xamarin.Android.Arch.Core.Common:\[1.1.1.1,)
```
    - metadata:
        id: "NuGet::Xamarin.Android.Arch.Core.Common:1.1.1.3"
      curations: []
```

Can you please check that?

**Comment by sschuberth** [comment]
> ivy-style version matcher introduction: https://docs.gradle.org/current/userguide/single_versions.html

I believe the better reference is https://ant.apache.org/ivy/history/2.1.0/settings/version-matchers.html.

**Comment by sschuberth** [comment]
> version range [1.1.1,)

For this issue, our underlying [vdurmont/semver4j](https://github.com/vdurmont/semver4j) library is known to have [some (similar) issues](https://github.com/vdurmont/semver4j/issues/67). Unfortunately, that library does not seem to be actively maintained anymore.

There's a maintained fork / copy at [semver4j/semver4j](https://github.com/semver4j/semver4j) that fixes some of the issues and has a significantly changes API. However, while [porting to it](https://github.com/oss-review-toolkit/ort/pull/6154) I discovered several other issues.

TL;DR, don't expect this to get fixed very soon.

> version range [1.1.1.1,)

That's actually something I would expected to get fixed at all. Such a four-digit version cannot easily be extended to a Semver (like a two-digit version), and [people actually expect it to not be accepted](https://github.com/vdurmont/semver4j/issues/60).

**Comment by mawl** [comment]
After further testing another bug hits the stage:

For transitive dependencies, curations with prefix matcher version (f.e.g 1.+) doesn't work either:

You can test it with:

`<PackageReference Include="Xamarin.Android.Arch.Core.Common" Version="1.1.1.3" />`

It has a [dependency](https://www.nuget.org/packages/Xamarin.Android.Arch.Core.Common/1.1.1.3#dependencies-body-tab) on:
```
- id: "NuGet::Xamarin.Android.Support.Annotations:28.+"
  curations:
    comment: >-
      https://go.microsoft.com/fwlink/?linkid=865381 contains MIT License
declared_license_mapping:
      "https://go.microsoft.com/fwlink/?linkid=865381": "MIT"
```
version matcher has to be removed for the curation to work here:
```
- id: "NuGet::Xamarin.Android.Support.Annotations"
  curations:
    comment: >-
      https://go.microsoft.com/fwlink/?linkid=865381 contains MIT License
declared_license_mapping:
      "https://go.microsoft.com/fwlink/?linkid=865381": "MIT"
```

**Comment by sschuberth** [comment]
> "https://go.microsoft.com/fwlink/?linkid=865381": "MIT"

As an unrelated side-note, mappings like these are a bit risky as the page's contents could change any time, mentioning a different license than MIT. I'm playing with https://github.com/oss-review-toolkit/ort/pull/5259 to resolve this.

**Comment by sschuberth** [comment]
Related: https://github.com/oss-review-toolkit/ort/pull/6922

**Comment by sschuberth** [comment]
> TL;DR, don't expect this to get fixed very soon.

By now we've migrated to [semver4j/semver4j](https://github.com/semver4j/semver4j), which also had several bug fix releases. Please reopen (and provide a test case) if there is anything left to do.
